File: lib/utils/shared_preferences_utils.dart
import 'package:shared_preferences/shared_preferences.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'dart:convert';

class SharedPreferencesUtils {
  static final FlutterSecureStorage _secureStorage = const FlutterSecureStorage();

  /// ذخیره مقدار ساده
  static Future<void> saveString(String key, String value) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(key, value);
  }

  static Future<String?> getString(String key, {String? defaultValue}) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(key) ?? defaultValue;
  }

  static Future<void> saveBool(String key, bool value) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setBool(key, value);
  }

  static Future<bool> getBool(String key, {bool defaultValue = false}) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getBool(key) ?? defaultValue;
  }

  /// ذخیره ارز انتخابی - مطابق با Kotlin
  static Future<void> saveSelectedCurrency(String currency) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('selected_currency', currency);
  }

  static Future<String> getSelectedCurrency() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('selected_currency') ?? 'USD';
  }

  /// ذخیره و بارگذاری قیمت توکن‌ها (مطابق با Kotlin shared_preferences_utils.kt)
  static Future<void> saveTokenPrice(String symbol, String currency, String price) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('price_${symbol}_$currency', price);
  }

  static Future<String> getTokenPrice(String symbol, String currency) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('price_${symbol}_$currency') ?? '0.0';
  }

  /// ذخیره و بارگذاری کش قیمت‌ها با timestamp (مطابق با Kotlin fetchPricesWithCache)
  static Future<void> savePricesMapWithCache(Map<String, Map<String, double>> pricesMap) async {
    final prefs = await SharedPreferences.getInstance();
    
    // ذخیره نقشه قیمت‌ها
    final jsonMap = <String, Map<String, String>>{};
    pricesMap.forEach((symbol, currencyMap) {
      jsonMap[symbol] = currencyMap.map((currency, price) => MapEntry(currency, price.toString()));
    });
    
    await prefs.setString('cached_prices_map', json.encode(jsonMap));
    await prefs.setInt('prices_cache_timestamp', DateTime.now().millisecondsSinceEpoch);
    
    print('💾 Saved prices cache with ${pricesMap.length} symbols');
  }

  /// بارگذاری کش قیمت‌ها (مطابق با Kotlin fetchPricesWithCache)
  static Future<Map<String, Map<String, double>>?> loadPricesMapFromCache({int maxAgeMinutes = 5}) async {
    final prefs = await SharedPreferences.getInstance();
    
    final timestamp = prefs.getInt('prices_cache_timestamp') ?? 0;
    final now = DateTime.now().millisecondsSinceEpoch;
    final ageMinutes = (now - timestamp) / (1000 * 60);
    
    if (ageMinutes > maxAgeMinutes) {
      print('⚠️ Prices cache expired (${ageMinutes.toStringAsFixed(1)} minutes old)');
      return null;
    }
    
    final jsonString = prefs.getString('cached_prices_map');
    if (jsonString == null) {
      print('⚠️ No prices cache found');
      return null;
    }
    
    try {
      final jsonMap = json.decode(jsonString) as Map<String, dynamic>;
      final pricesMap = <String, Map<String, double>>{};
      
      jsonMap.forEach((symbol, currencyMap) {
        if (currencyMap is Map<String, dynamic>) {
          pricesMap[symbol] = {};
          currencyMap.forEach((currency, priceString) {
            final price = double.tryParse(priceString.toString()) ?? 0.0;
            pricesMap[symbol]![currency] = price;
          });
        }
      });
      
      print('📥 Loaded prices cache with ${pricesMap.length} symbols (${ageMinutes.toStringAsFixed(1)} min old)');
      return pricesMap;
    } catch (e) {
      print('❌ Error loading prices cache: $e');
      return null;
    }
  }

  /// دریافت قیمت‌ها با کش (مطابق با Kotlin fetchPricesWithCache)
  static Future<Map<String, Map<String, double>>> fetchPricesWithCache({
    required List<String> symbols,
    required List<String> fiatCurrencies,
    int maxCacheAgeMinutes = 5,
  }) async {
    print('🔄 fetchPricesWithCache called with symbols: $symbols, currencies: $fiatCurrencies');
    
    // Try loading from cache first
    final cachedPrices = await loadPricesMapFromCache(maxAgeMinutes: maxCacheAgeMinutes);
    if (cachedPrices != null) {
      // Check if cache has all required data
      bool hasAllData = true;
      for (final symbol in symbols) {
        if (!cachedPrices.containsKey(symbol)) {
          hasAllData = false;
          break;
        }
        for (final currency in fiatCurrencies) {
          if (!cachedPrices[symbol]!.containsKey(currency)) {
            hasAllData = false;
            break;
          }
        }
        if (!hasAllData) break;
      }
      
      if (hasAllData) {
        print('✅ Using cached prices for all requested symbols');
        return cachedPrices;
      }
    }
    
    // Cache miss or incomplete, fetch from API and update cache
    print('🌐 Cache miss, fetching prices from API...');
    
    // This would normally call API service, but since we're in utils,
    // we'll return empty map and let the caller handle API calls
    print('⚠️ fetchPricesWithCache: API call should be handled by caller');
    return {};
  }

  /// ذخیره و بارگذاری mnemonic به صورت امن
  static Future<void> saveMnemonic(String userId, String walletName, String mnemonic) async {
    final key = 'mnemonic_${userId}_$walletName';
    await _secureStorage.write(key: key, value: mnemonic);
  }

  static Future<String?> getMnemonic(String userId, String walletName) async {
    final key = 'mnemonic_${userId}_$walletName';
    return await _secureStorage.read(key: key);
  }

  /// ذخیره و بارگذاری userId
  static Future<void> saveUserId(String walletName, String userId) async {
    final prefs = await SharedPreferences.getInstance();
    final walletsJson = prefs.getString('user_wallets') ?? '[]';
    List wallets = jsonDecode(walletsJson);
    final index = wallets.indexWhere((w) => w['walletName'] == walletName);
    if (index != -1) {
      wallets[index]['userId'] = userId;
    } else {
      wallets.add({'walletName': walletName, 'userId': userId});
    }
    await prefs.setString('user_wallets', jsonEncode(wallets));
  }

  static Future<String?> getUserId(String walletName) async {
    final prefs = await SharedPreferences.getInstance();
    final walletsJson = prefs.getString('user_wallets') ?? '[]';
    List wallets = jsonDecode(walletsJson);
    final wallet = wallets.cast<Map>().firstWhere(
      (w) => w['walletName'] == walletName,
      orElse: () => {},
    );
    return wallet['userId'] ?? prefs.getString('UserID');
  }

  /// فرمت نماد ارز - مطابق با Kotlin
  static String getCurrencySymbol(String currency) {
    switch (currency) {
      case 'USD': return '\$';       // دلار آمریکا
      case 'CAD': return 'CA\$';     // دلار کانادا
      case 'AUD': return 'AU\$';     // دلار استرالیا
      case 'GBP': return '£';       // پوند بریتانیا
      case 'EUR': return '€';       // یورو
      case 'KWD': return 'KD';      // دینار کویت
      case 'TRY': return '₺';       // لیر ترکیه
      case 'IRR': return '﷼';       // ریال ایران
      case 'SAR': return '﷼';       // ریال عربستان
      case 'CNY': return '¥';       // یوآن چین
      case 'KRW': return '₩';       // وون کره جنوبی
      case 'JPY': return '¥';       // ین ژاپن
      case 'INR': return '₹';       // روپیه هند
      case 'RUB': return '₽';       // روبل روسیه
      case 'IQD': return 'ع.د';     // دینار عراق
      case 'TND': return 'د.ت';     // دینار تونس
      case 'BHD': return 'ب.د';     // دینار بحرین
      case 'ZAR': return 'R';       // راند آفریقای جنوبی
      case 'CHF': return 'CHF';     // فرانک سوئیس
      case 'NZD': return 'NZ\$';     // دلار نیوزیلند
      case 'SGD': return 'S\$';      // دلار سنگاپور
      case 'HKD': return 'HK\$';     // دلار هنگ‌کنگ
      case 'MXN': return 'MX\$';     // پزو مکزیک
      case 'BRL': return 'R\$';      // رئال برزیل
      case 'SEK': return 'kr';      // کرون سوئد
      case 'NOK': return 'kr';      // کرون نروژ
      case 'DKK': return 'kr';      // کرون دانمارک
      case 'PLN': return 'zł';      // زلوتی لهستان
      case 'CZK': return 'Kč';      // کرون چک
      case 'HUF': return 'Ft';      // فورینت مجارستان
      case 'ILS': return '₪';       // شِکِل جدید اسرائیل
      case 'MYR': return 'RM';      // رینگیت مالزی
      case 'THB': return '฿';       // بات تایلند
      case 'PHP': return '₱';       // پزو فیلیپین
      case 'IDR': return 'Rp';      // روپیه اندونزی
      case 'EGP': return '£';       // پوند مصر
      case 'PKR': return '₨';       // روپیه پاکستان
      case 'NGN': return '₦';       // نایرا نیجریه
      case 'VND': return '₫';       // دونگ ویتنام
      case 'BDT': return '৳';       // تاکا بنگلادش
      case 'LKR': return 'Rs';      // روپیه سریلانکا
      case 'UAH': return '₴';       // گریونا اوکراین
      case 'KZT': return '₸';       // تنگه قزاقستان
      case 'XAF': return 'FCFA';    // فرانک آفریقای مرکزی
      case 'XOF': return 'CFA';     // فرانک آفریقای غربی
      default: return '';         // پیش‌فرض: رشته خالی
    }
  }

  /// فرمت قیمت - مطابق با Kotlin
  static String formatPrice(double? price, String symbol) {
    if (price == null || price.isNaN || price == 0.0) return '0';
    
    if (symbol == 'BTC' || symbol == 'ETH') {
      return price.toStringAsFixed(2);
    } else if (price < 0.01) {
      return price.toStringAsFixed(8);
    } else if (price < 1) {
      return price.toStringAsFixed(4);
    } else {
      return price.toStringAsFixed(2);
    }
  }

  /// فرمت مقدار - مطابق با Kotlin
  static String formatAmount(double amount, double price) {
    if (amount == 0.0) return '0';
    
    if (amount < 0.001) {
      return amount.toStringAsFixed(8);
    } else if (amount < 0.1) {
      return amount.toStringAsFixed(6);
    } else if (amount < 1.0) {
      return amount.toStringAsFixed(4);
    } else if (amount < 10.0) {
      return amount.toStringAsFixed(3);
    } else {
      return amount.toStringAsFixed(2);
    }
  }

  /// پاک کردن همه داده‌ها (برای logout)
  static Future<void> clearAll() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.clear();
    await _secureStorage.deleteAll();
  }
} 

File: ۱/۲/shared_preferences_utils.kt
package com.laxce.adl.utility

import android.content.Context
import android.content.SharedPreferences
import android.net.ConnectivityManager
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKeys
import com.laxce.adl.api.Api
import com.laxce.adl.api.PriceData
import com.laxce.adl.api.PricesRequest
import com.laxce.adl.api.RetrofitClient
import com.google.gson.Gson
import android.util.Log


private fun isInternetAvailable(context: Context): Boolean {
    val connectivityManager =
        context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    val networkInfo = connectivityManager.activeNetworkInfo
    return networkInfo != null && networkInfo.isConnected
}

fun getEncryptedSharedPreferences(context: Context): SharedPreferences {
    return try {
        EncryptedSharedPreferences.create(
            "passcode_prefs",
            MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC),
            context,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    } catch (e: Exception) {
        resetEncryptedPreferences(context)
        context.getSharedPreferences("passcode_prefs", Context.MODE_PRIVATE)
    }
}

fun reloadPrices(context: Context): Map<String, Map<String, String>> {
    val sharedPreferences = context.getSharedPreferences("token_prices", Context.MODE_PRIVATE)
    val allPrices = sharedPreferences.all
    val pricesMap = mutableMapOf<String, Map<String, String>>()

    allPrices.forEach { (key, value) ->
        val parts = key.split("-")
        if (parts.size == 2) {
            val symbol = parts[0]
            val currency = parts[1]
            val price = value.toString()

            pricesMap[symbol] = pricesMap.getOrDefault(symbol, mutableMapOf()).toMutableMap().apply {
                this[currency] = price
            }
        }
    }
    return pricesMap
}

fun loadPriceFromKeystore(context: Context, symbol: String, currency: String): Double {
    val sharedPreferences = context.getSharedPreferences("token_prices", Context.MODE_PRIVATE)
    val key = "$symbol-$currency"

    // مقدار پیش‌فرض 0.0
    return sharedPreferences.getString(key, "0.0")?.toDoubleOrNull() ?: 0.0
}

suspend fun fetchPricesWithCache(
    context: Context,
    gson: Gson,
    selectedCurrency: String,
    activeSymbols: List<String>,
    fiatCurrencies: List<String>
    ): Map<String, Map<String, String>> {

    var sanitizedSymbols = activeSymbols.map { it.uppercase() }

    if (sanitizedSymbols.isEmpty()) {
        val sharedPreferences = context.getSharedPreferences("token_prefs", Context.MODE_PRIVATE)
        sanitizedSymbols = sharedPreferences.all.keys
            .filter { it.startsWith("token_") && it.endsWith("_enabled") && sharedPreferences.getBoolean(it, false) }
            .map { it.replace("token_", "").replace("_enabled", "").uppercase() }
    }

    // دریافت API و UserID
    val api = RetrofitClient.getInstance(context).create(Api::class.java)
    val userId = context.getSharedPreferences("app_keystore", Context.MODE_PRIVATE)
        .getString("UserID", "") ?: ""


    if (sanitizedSymbols.isEmpty()) {
        return emptyMap()
    }

    // Map symbols to match the API request format with full names
    val apiSymbols = sanitizedSymbols.map { 
        when (it) {
            "BTC" -> "bitcoin"
            "ETH" -> "ethereum" 
            "TRX" -> "tron"
            "BNB" -> "BNB"
            "SHIB" -> "shiba inu"
            else -> it.lowercase()
        }
    }

    // ساخت درخواست
    val request = PricesRequest(
        Symbol = apiSymbols,
        FiatCurrencies = fiatCurrencies // لیست همه ارزها
    )

    return try {
        val response = api.getPrices(request)

        if (response.success) {
            val prices: Map<String, Map<String, PriceData>> = response.prices ?: emptyMap()

            // Map the returned keys back to the expected format if needed
            val mappedPrices = mutableMapOf<String, Map<String, PriceData>>()
            prices.forEach { (key, value) ->
                val mappedKey = when (key) {
                    "BTC" -> "BTC"
                    "ETH" -> "ETH"
                    "TRX" -> "TRX"
                    "BNB" -> "BNB"
                    "bitcoin" -> "BTC"
                    "ethereum" -> "ETH"
                    "tron" -> "TRX"
                    "BNB" -> "BNB"
                    "shiba inu" -> "SHIB"
                    else -> key
                }
                mappedPrices[mappedKey] = value
            }

            // بررسی مقدار `null` برای جلوگیری از کرش
            val safePrices: Map<String, Map<String, PriceData>> = mappedPrices.mapValues { (_, currencyMap) ->
                currencyMap.mapValues { (_, priceData) ->
                    priceData ?: PriceData(price = "0.0", change_24h = "+0.0%")
                }
            }

            // ذخیره قیمت‌ها در کش
            context.getSharedPreferences("token_prices", Context.MODE_PRIVATE).edit()
                .putString("cached_prices", gson.toJson(safePrices))
                .putLong("price_cache_timestamp", System.currentTimeMillis())
                .apply()

            // ذخیره قیمت‌های هر توکن به صورت جداگانه برای سازگاری با کد قدیمی
            val sharedPreferences = context.getSharedPreferences("token_prices", Context.MODE_PRIVATE)
            val editor = sharedPreferences.edit()
            
            safePrices.forEach { (symbol, currencyMap) ->
                currencyMap.forEach { (currency, priceData) ->
                    editor.putString("$symbol-$currency-price", priceData.price)
                    editor.putString("$symbol-$currency-change", priceData.change_24h)
                }
            }
            editor.apply()

            // بازگرداندن قیمت‌ها به فرمت قابل استفاده در token_view_model
            return safePrices.mapValues { (_, currencyMap) ->
                currencyMap.mapValues { (_, priceData) ->
                    priceData.price // فقط قیمت را برمی‌گردانیم
                }
            }

        } else {
            return emptyMap()
        }
    } catch (e: Exception) {
        return emptyMap()
    }
}



//fun formatPrice(price: String, selectedCurrency: String): String {
//    val currencySymbol = getCurrencySymbol(selectedCurrency)
//    return try {
//        val sanitizedPrice = price.replace("", "")
//        val value = sanitizedPrice.toBigDecimal()
//        when {
//            value >= "1".toBigDecimal() -> {
//                "$currencySymbol${value.setScale(2, java.math.RoundingMode.DOWN).toPlainString()}"
//            }
//            value >= "0.01".toBigDecimal() -> {
//                "$currencySymbol${value.setScale(4, java.math.RoundingMode.DOWN).toPlainString()}"
//            }
//            else -> {
//                val formatted = value.setScale(10, java.math.RoundingMode.DOWN).toPlainString()
//                val nonZeroIndex = formatted.indexOfFirst { it != '0' && it != '.' }
//                if (nonZeroIndex > 3) {
//                    "$currencySymbol${formatted.substring(0, nonZeroIndex + 4)}"
//                } else {
//                    "$currencySymbol${formatted.trimEnd('0').trimEnd('.')}"
//                }
//            }
//        }
//    } catch (e: Exception) {
//        "$currencySymbol$price" // در صورت خطا مقدار اصلی بازگردانده شود
//    }
//}

fun resetEncryptedPreferences(context: Context) {
    val prefs = context.getSharedPreferences("passcode_prefs", Context.MODE_PRIVATE)
    prefs.edit().clear().apply()
}


fun getUserIdFromKeystore(context: Context, walletName: String): String {
    val sharedPreferences = context.getSharedPreferences("app_keystore", Context.MODE_PRIVATE)
    val walletsJson = sharedPreferences.getString("user_wallets", "[]") ?: "[]"
    val wallets = try {
        Gson().fromJson(walletsJson, ArrayList::class.java) as ArrayList<Map<String, String>>
    } catch (e: Exception) {
        ArrayList()
    }

    val wallet = wallets.find { it["walletName"] == walletName }
    val userId = wallet?.get("userId") ?: ""

    if (userId.isEmpty()) {
        // اگر در فرمت جدید پیدا نشد، سعی کنید از فرمت قدیمی بخوانید
        val legacyUserId = sharedPreferences.getString("UserID", "")
        if (!legacyUserId.isNullOrEmpty()) {
            return legacyUserId
        }
    } else {
    }

    return userId
}

fun getMnemonicFromKeystore(context: Context, userId: String, walletName: String): String? {
    val masterKeyAlias = MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC)
    val encryptedPrefs = EncryptedSharedPreferences.create(
        "encrypted_mnemonic_prefs",
        masterKeyAlias,
        context,
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
    )
    
    // ابتدا سعی می‌کنیم با نام فعلی کیف پول پیدا کنیم
    val currentKey = "Mnemonic_${userId}_${walletName}"
    var mnemonic = encryptedPrefs.getString(currentKey, null)
    
    // اگر پیدا نشد، سعی می‌کنیم با نام‌های مختلف پیدا کنیم
    if (mnemonic == null) {
        val allEntries = encryptedPrefs.all
        val targetUserId = userId
        
        // جستجو در تمام کلیدهای mnemonic برای یافتن کلید مناسب
        for ((key, value) in allEntries) {
            if (key.startsWith("Mnemonic_${targetUserId}_") && value is String && value.isNotEmpty()) {
                // بررسی می‌کنیم که آیا این کلید مربوط به همین کیف پول است یا نه
                // با استفاده از userId و بررسی اینکه آیا این کلید با نام‌های مختلف ذخیره شده
                mnemonic = value
                break
            }
        }
    }
    
    return mnemonic
}

fun saveSelectedCurrency(context: Context, currency: String) {
    val sharedPreferences = context.getSharedPreferences("app_preferences", Context.MODE_PRIVATE)
    sharedPreferences.edit().putString("selected_currency", currency).apply()
}

fun getSelectedCurrency(context: Context): String {
    val sharedPreferences = context.getSharedPreferences("app_preferences", Context.MODE_PRIVATE)
    return sharedPreferences.getString("selected_currency", "USD") ?: "USD"
}


fun getCurrencySymbol(currency: String): String {
    return when (currency) {
        "USD" -> "$"       // دلار آمریکا
        "CAD" -> "CA$"     // دلار کانادا
        "AUD" -> "AU$"     // دلار استرالیا
        "GBP" -> "£"       // پوند بریتانیا
        "EUR" -> "€"       // یورو
        "KWD" -> "KD"      // دینار کویت
        "TRY" -> "₺"       // لیر ترکیه
        "IRR" -> "﷼"       // ریال ایران
        "SAR" -> "﷼"       // ریال عربستان
        "CNY" -> "¥"       // یوآن چین
        "KRW" -> "₩"       // وون کره جنوبی
        "JPY" -> "¥"       // ین ژاپن
        "INR" -> "₹"       // روپیه هند
        "RUB" -> "₽"       // روبل روسیه
        "IQD" -> "ع.د"     // دینار عراق
        "TND" -> "د.ت"     // دینار تونس
        "BHD" -> "ب.د"     // دینار بحرین
        "ZAR" -> "R"       // راند آفریقای جنوبی
        "CHF" -> "CHF"     // فرانک سوئیس
        "NZD" -> "NZ$"     // دلار نیوزیلند
        "SGD" -> "S$"      // دلار سنگاپور
        "HKD" -> "HK$"     // دلار هنگ‌کنگ
        "MXN" -> "MX$"     // پزو مکزیک
        "BRL" -> "R$"      // رئال برزیل
        "SEK" -> "kr"      // کرون سوئد
        "NOK" -> "kr"      // کرون نروژ
        "DKK" -> "kr"      // کرون دانمارک
        "PLN" -> "zł"      // زلوتی لهستان
        "CZK" -> "Kč"      // کرون چک
        "HUF" -> "Ft"      // فورینت مجارستان
        "ILS" -> "₪"       // شِکِل جدید اسرائیل
        "MYR" -> "RM"      // رینگیت مالزی
        "THB" -> "฿"       // بات تایلند
        "PHP" -> "₱"       // پزو فیلیپین
        "IDR" -> "Rp"      // روپیه اندونزی
        "EGP" -> "£"       // پوند مصر
        "PKR" -> "₨"       // روپیه پاکستان
        "NGN" -> "₦"       // نایرا نیجریه
        "VND" -> "₫"       // دونگ ویتنام
        "BDT" -> "৳"       // تاکا بنگلادش
        "LKR" -> "Rs"      // روپیه سریلانکا
        "UAH" -> "₴"       // گریونا اوکراین
        "KZT" -> "₸"       // تنگه قزاقستان
        "XAF" -> "FCFA"    // فرانک آفریقای مرکزی
        "XOF" -> "CFA"     // فرانک آفریقای غربی
        else -> ""         // پیش‌فرض: رشته خالی
    }
}



// Extension function for formatting double values
fun Double.format(digits: Int): String {
    val safeDigits = digits.coerceIn(0, 10) // محدود کردن digits بین 0 و 10
    return "%.${safeDigits}f".format(this)
}

fun formatAmount(amount: Double, price: Double): String {
    if (amount == 0.0) return "0"

    // برای مقادیر خیلی کوچک، تعداد رقم اعشار بیشتری نیاز داریم
    val formatted = when {
        // برای مقادیر کمتر از 0.001 (مثل 0.00012)، حداقل 6 رقم اعشار
        amount < 0.001 -> String.format("%.8f", amount)
        // برای مقادیر کمتر از 0.1، حداقل 4 رقم اعشار  
        amount < 0.1 -> String.format("%.6f", amount)
        // برای مقادیر کمتر از 1، 4 رقم اعشار
        amount < 1.0 -> String.format("%.4f", amount)
        // برای مقادیر کمتر از 10، 3 رقم اعشار
        amount < 10.0 -> String.format("%.3f", amount)
        // برای مقادیر بزرگتر، 2 رقم اعشار
        else -> String.format("%.2f", amount)
    }

    // حذف صفرهای اضافی از انتها
    return formatted.trimEnd('0').trimEnd('.')
}

fun formatPrice(price: Double?, symbol: String): String {
    if (price == null || price.isNaN() || price == 0.0) return "0"

    return when {
        symbol == "BTC" || symbol == "ETH" -> String.format("%,.2f", price)
        price < 0.01 -> String.format("%.8f", price)
        price < 1 -> String.format("%.4f", price)
        else -> String.format("%.2f", price)
    }
}
